apps/registry/builder/index.ts

```diff
@@ -63,41 +76,101 @@ export const PIRANHA_LANGUAGES = [
 	"scala",
 ] as const;
 
-const piranhaLanguageSchema = S.union(
-	...PIRANHA_LANGUAGES.map((language) => S.literal(language)),
-);
-
-const codemodConfigSchema = S.union(
-	S.struct({
-		schemaVersion: S.literal("1.0.0"),
-		engine: S.literal("piranha"),
-		language: piranhaLanguageSchema,
-		arguments: optionalArgumentsSchema,
-	}),
-	S.struct({
-		schemaVersion: S.literal("1.0.0"),
-		engine: S.literal("jscodeshift"),
-		arguments: optionalArgumentsSchema,
-	}),
-	S.struct({
-		schemaVersion: S.literal("1.0.0"),
-		engine: S.literal("ts-morph"),
-		arguments: optionalArgumentsSchema,
-	}),
-	S.struct({
-		schemaVersion: S.literal("1.0.0"),
-		engine: S.literal("filemod"),
-		arguments: optionalArgumentsSchema,
-	}),
-	S.struct({
-		schemaVersion: S.literal("1.0.0"),
-		engine: S.literal("recipe"),
-		names: S.array(S.string),
-		arguments: optionalArgumentsSchema,
+// Source: https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+const semVerRegex =
+	/^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/;
+
+const knownEngines = [
+	literal("jscodeshift"),
+	literal("repomod-engine"),
+	literal("filemod"),
+	literal("ts-morph"),
+	literal("ast-grep"),
+];
+
+const versionValidator = union([
+	// react < 18.0.2 (preferred)
+	string([regex(semVerRegex)]),
+	// react < 18 (for example, when no latest version of a given major is out yet)
+	coerce(number(), (input) => Number(input)),
+]);
+
+const configJsonBaseSchema = object({
```

Comment:
can we reuse `configJsonBaseSchema`? currently its duplicated in cli and registry

Reply:
To be honest, in my opinion, this builder functionality is going to be partly moved to our backend, where we will re-upload registry.tar.gz every time a new public codemod is pushed from the CLI, and registry will not be responsible for uploading files to s3 anymore. This builder is subject to deprecation, so I think no need to find a solution for that issue right now.